doc/neps/nep-0018-array-function-protocol.rst

```diff
@@ -187,27 +193,60 @@ of these may decide that, given the available inputs, they are unable to
 determine the correct result. How do we call the right one? If several
 are valid then which has precedence?
 
-The rules for dispatch with ``__array_function__`` match those for
-``__array_ufunc__`` (see
+For the most part, the rules for dispatch with ``__array_function__``
+match those for ``__array_ufunc__`` (see
 `NEP-13 <http://www.numpy.org/neps/nep-0013-ufunc-overrides.html>`_).
 In particular:
 
 -  NumPy will gather implementations of ``__array_function__`` from all
    specified inputs and call them in order: subclasses before
-   superclasses, and otherwise left to right. Note that in some edge cases,
-   this differs slightly from the
+   superclasses, and otherwise left to right. Note that in some edge cases
+   involving subclasses, this differs slightly from the
    `current behavior <https://bugs.python.org/issue30140>`_ of Python.
 -  Implementations of ``__array_function__`` indicate that they can
    handle the operation by returning any value other than
    ``NotImplemented``.
 -  If all ``__array_function__`` methods return ``NotImplemented``,
    NumPy will raise ``TypeError``.
 
+One deviation from the current behavior of ``__array_ufunc__`` is that NumPy
+will only call ``__array_function__`` on the *first* argument of each unique
+type. This matches Python's
+`rule for calling reflected methods <https://docs.python.org/3/reference/datamodel.html#object.__ror__>`_,
+and this ensures that checking overloads has acceptable performance even when
+there are a large number of overloaded arguments.
+
+Special handling of ``numpy.ndarray``
+'''''''''''''''''''''''''''''''''''''
+
+The use cases for subclasses with ``__array_function__`` are the same as those
+with ``__array_ufunc__``, so ``numpy.ndarray`` should also define a
+``__array_function__`` method mirroring ``ndarray.__array_ufunc__``:
+
+.. code:: python
+
+    def __array_function__(self, func, types, args, kwargs):
+        # Cannot handle items that have __array_function__ other than our own.
+        for t in types:
+            if (hasattr(t, '__array_function__') and
+                    t.__array_function__ is not ndarray.__array_function__):
+                return NotImplemented
+
+        # Arguments contain no overrides, so we can safely call the
+        # overloaded function again.
+        return func(*args, **kwargs)
+
+To avoid recursion, the dispatch rules for ``__array_function__`` need also
+the same special case they have for ``__array_ufunc__``: any arguments with an
```

Comment:
*OR* the non-wrapped function can be called. That would give a substantial speed advantage. (For `__array_ufunc__`, part of #8892)